### PR TITLE
Implement slashing protocol

### PR DESCRIPTION
**Pull request by linked0**
With this PR, we can add information about validators being slashed into the block. After the validators being slashed, the UTXOs of the validators are split into the validators' address and the `CommonsBudget` address.

This does *NOT* contain the validation check for the information yet. But it will be implemented soon. The feature of PR #1421 is the need for the next step, which is to maintain un-enrolled validators in the `ValidatorSet`.

Part of #1076

### CODE REVIEW

File: source/agora/consensus/EnrollmentManager.d [hunk 1]
```diff
@@ -565,12 +565,13 @@ public class EnrollmentManager
         foreach (const ref key; keys)
         {
             const preimage = this.validator_set.getPreimageAt(key, height);
-            // this should not happen. validators which didn't reveal the
-            // preimage should not be in the active validator set 'keys'.
+            // We could have a misbehaving validator that does not reveal
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d:705]: The way I envisioned it is that we would simply return the preimage root that's stored in the block header.

This function is always called with a block height in the past.

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: But I guess the next commits do that. I'm still reviewing.

File: source/agora/consensus/data/Block.d [hunk 2]
```diff
@@ -59,6 +59,12 @@ public struct BlockHeader
     /// Enrolled validators
     public Enrollment[] enrollments;
 
+    /// Hash of random seed of the preimages for this this height
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/Block.d:62]: in the commit message: whil => while

File: source/agora/node/Ledger.d [hunk 3]
```diff
@@ -518,8 +522,8 @@ public class Ledger
 
     public string validateSlashingData (in ConsensusData data) @safe
     {
-        return this.slash_man.isInvalidPreimageRootReason(this.getBlockHeight(),
-                data.missing_validators);
+        return this.slash_man.isInvalidPreimageRootReason(
+                this.last_block.header.height, data.missing_validators);
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Is this change just an optimization? Or is the height passed here different now?

File: source/agora/consensus/EnrollmentManager.d [hunk 4]
```diff
@@ -924,6 +924,20 @@ public class EnrollmentManager
...
+
+        Params:
+            enroll_hash = key for the validator to remove
+
+    ***************************************************************************/
+
+    public void removeValidator (in Hash enroll_hash) @trusted
+    {
+        this.validator_set.remove(enroll_hash);
+    }
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d:1111]: I think we're going with the other approach which is just tagging the disabled validators, and not removing them from the set. This is to support what was implemented in https://github.com/bpfkorea/agora/pull/1421/commits

File: source/agora/consensus/SlashPolicy.d [hunk 5]
```diff
@@ -113,6 +113,36 @@ public class SlashPolicy
...
+
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys))
+        {
+            log.fatal("Could not retrieve enrollments");
+            assert(0);
+        }
+
+        foreach (idx; missing_validators)
+        {
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d:141]: This isn't actually doing anything? The array is empty.

File: source/agora/node/Ledger.d (hunk 3 above)
Comment (linked0) [reply source/agora/node/Ledger.d]: I changed it unintentionally. There's no difference between before and after. I'll restore it.

File: source/agora/consensus/state/UTXOSet.d [hunk 6]
```diff
@@ -49,18 +49,20 @@ abstract class UTXOCache
         Params:
             tx = the transaction
             height = local height of the block
+            slashed = whether the transaction is for slashing
```
Comment (AndrejMitrovic) [inline source/agora/consensus/state/UTXOSet.d:52]: typo in commit message: chian

File: source/agora/consensus/EnrollmentManager.d (hunk 4 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:1111]: > I think we're going with the other approach which is just tagging the disabled validators, and not removing them from the set. This is to support what was implemented in https://github.com/bpfkorea/agora/pull/1421/commits

Yes, right. The `remove` function will just change the flag of 'active' after the commit being merged.

File: source/agora/consensus/data/Block.d (hunk 2 above)
Comment (linked0) [reply source/agora/consensus/data/Block.d:62]: > in the commit message: whil => while

corrected.

File: source/agora/consensus/state/UTXOSet.d (hunk 6 above)
Comment (linked0) [reply source/agora/consensus/state/UTXOSet.d:52]: > typo in commit message: chian

corrected.

File: source/agora/consensus/EnrollmentManager.d (hunk 4 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:1111]: Added `unenrollValidator` function instead of `removeValidator` and new function called `unenroll` into the `ValidatorSet`.

File: source/agora/consensus/SlashPolicy.d (hunk 5 above)
Comment (linked0) [reply source/agora/consensus/SlashPolicy.d:141]: > This isn't actually doing anything? The array is empty.

The `missing_validators` which is in fact `missing_validators` in `BlockHeader` is passed from the Ledger when creating a new Block. Can you explain why do you think it's empty?

File: source/agora/consensus/EnrollmentManager.d (hunk 1 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: > The way I envisioned it is that we would simply return the preimage root that's stored in the block header.
> 
> This function is always called with a block height in the past.

I will think more about that.

File: source/agora/consensus/SlashPolicy.d (hunk 5 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/SlashPolicy.d:141]: ahh I misread the code. I saw the other array set to zero.

File: source/agora/consensus/SlashPolicy.d [hunk 7]
```diff
@@ -113,6 +113,36 @@ public class SlashPolicy
...
+        by indices
+
+        Params:
+            validators_utxos = will contain the UTXOs ot the validators
+            missing_validators = indices of validators being slashed
+
+    ***************************************************************************/
+
+    public void getMissingValidatorsUTXOs (ref Hash[] validators_utxos,
+        const ref uint[] missing_validators) @safe nothrow
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: This doesn't need to be `ref`.

File: source/agora/consensus/EnrollmentManager.d (hunk 1 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: > The way I envisioned it is that we would simply return the preimage root that's stored in the block header.
> 
> This function is always called with a block height in the past.

I thought about it. 
The `random_seed` of a block header is for the previous height and the `rand_seed` from this `getRandomSeed` function which is used for the next block is for the current height. So we should generate a random seed again in this function, I think.

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: > which is used for the next block 

I thought this function was used for the quorum generator. When it generates quorums for height N it uses preimages of N-1. So it could just read the root from the N-1 block header.

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: I think the problem I am seeing here is that `getRandomSeed` is used by both the quorum generator and the slasher.

The quorum generator needs a fixed (externalized) random seed, which we are now storing in the block header.

So maybe there should be two functions? One is the `getCandidateRandomSeed` and the other `getExternalizedRandomSeed`. Do you see what I mean?

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: > > which is used for the next block
> 
> I thought this function was used for the quorum generator. When it generates quorums for height N it uses preimages of N-1. So it could just read the root from the N-1 block header.

I thought it the same way, but I realized that the random seed from the N-1 block header is generated with preimages of N-2 in rethinking. Do you mean that the random seed is generated with pre-images of current validators and newly enrolled validators?

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: > So maybe there should be two functions? One is the `getCandidateRandomSeed` and the other `getExternalizedRandomSeed`. Do you see what I mean?

I got it. I thought it right before. We can generate the random seed for the `getExternalizedRandomSeed` with pre-images of current validators and newly enrolled validators when validating a block, which is stored in a block header after.

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: So I took another look now.

When a block is externalized, we call `rebuildQuorumConfig(.., block.height)`. That function calls `getRandomSeed()`.

What I am worried about is the case where a validator reveals their preimage in the small time window just before a block is externalized. Their preimage might not be included in the block's header (because it's too late), but it might get included in some node's local validator set because a new block was not externalized yet. Then, when `rebuildQuorumConfig()` is called it's possible some nodes will use a different random seed based on their current internal preimage state.

That's why I suggest to use a separate function for the quorum generator. It should take the random seed in the block header (always). I was wrong about it taking the `N-1`'th preimages, it's the `N`'th preimages.

Does that make sense?

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: > That's why I suggest to use a separate function for the quorum generator. It should take the random seed in the block header (always). I was wrong about it taking the N-1'th preimages, it's the N'th preimages.
>
> Does that make sense?

Yes, that makes sense. And there is one more thing to be considered. Which pre-images are used for the `random seed` for the block header? There is a situation some pre-images are NOT revealed at the current height (I mean, the height of the currently externalized block) So, we need to generate the random seed with N-1'th preimages of the existing validator and N'th preimages of newly enrolled validators. What do you think of it?

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: > There is a situation some pre-images are NOT revealed at the current height (I mean, the height of the currently externalized block)

I mean that's fine? After the block is externalized we use that random seed for the quorum generator. We don't care what's in it, just that it was voted and externalized.

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: > I mean that's fine? After the block is externalized we use that random seed for the quorum generator. We don't care what's in it, just that it was voted and externalized.

The `random_seed` is not part of `ConsensusData`. So we should determine what pre-images should used for the `random_seed`, I think. (EDITED)

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: I am just talking about the quorum generator though. Not the nomination process.

Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: If this isn't making much sense to you, it's fine to keep it the way you wrote it.

I'll submit a separate issue for this later.

Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: I will think about what you mean again if you submit a new issue about that. Anyway, I just wanted to know how to generate the `random_seed` for a block header.
Thank you.

**Comment by linked0** [comment]
Ready for review.
Thanks.

File: source/agora/consensus/EnrollmentManager.d (hunk 1 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/EnrollmentManager.d:705]: > Anyway, I just wanted to know how to generate the random_seed for a block header.

Yes, that's fair. I was looking at the other use-case and not specifically your own. So it may have been off-topic for this PR. I apologize.

**Comment by AndrejMitrovic** [comment]
Unfortunately there is a conflict now. I hope it won't be hard to rebase?

**Comment by linked0** [comment]
> Unfortunately there is a conflict now. I hope it won't be hard to rebase?

I'm rebasing it now.

File: source/agora/consensus/EnrollmentManager.d (hunk 1 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d:705]: > > Anyway, I just wanted to know how to generate the random_seed for a block header.
> 
> Yes, that's fair. I was looking at the other use-case and not specifically your own. So it may have been off-topic for this PR. I apologize.

It's okay. There are some mistakes in expressing my opinion in English. Thanks.

**Comment by AndrejMitrovic** [comment]
Thank you for rebasing. I'm doing a review now.

File: source/agora/utils/PrettyPrinter.d [hunk 8]
```diff
@@ -355,7 +358,9 @@ private struct BlockHeaderFmt
 { utxo: 0xb20d...08eb, seed: 0xa050...2cb4, cycles: 20, sig: 0x052e...6b31 }
 { utxo: 0xdb39...2d85, seed: 0xdd1b...7bfa, cycles: 20, sig: 0x0e00...4fe2 }]
 Signature: 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000,
-Validators: [0]`;
+Validators: [0],
+RandomSeed: [0x0000...0000],
+MPV: []`;
```
Comment (AndrejMitrovic) [inline source/agora/utils/PrettyPrinter.d]: In the future I think we should name this just `MV` that stands for "missing validators".

File: source/agora/consensus/data/Block.d [hunk 9]
```diff
@@ -541,9 +552,16 @@ version (unittest)
             "7caa4fcffdc5c068a07532637cf5042ae39b7af418847385480e620e1395987")
     };
 
-    auto block = makeNewBlock(GenesisBlock, [Transaction.init], [enr_1, enr_2]);
+    auto primage_root = Hash("0x47c993d409aa7d77651ecaa5a5d29e47a7aee609c7" ~
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/Block.d]: `primage_root` => `preimage_root`

File: source/agora/node/Ledger.d [hunk 10]
```diff
@@ -229,7 +229,10 @@ public class Ledger
     public bool onExternalized (ConsensusData data)
         @trusted
     {
-        auto block = makeNewBlock(this.last_block, data.tx_set, data.enrolls);
+        Hash random_seed = slash_man.getRandomSeed(this.getBlockHeight());
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: One change can be made here. Take into account the `missing_validators` in the ConsensusData, and make sure their preimage is not included when calculating the random seed inside of `getRandomSeed`.

The reason why this is needed is because there can be a timing issue. A block could be externalized around the same time as another slow validator sends a preimage. So that validator could be inside of `missing_validators`, but the preimage could have been added just a second before externalization.

This is the issue I talked about before.

File: source/agora/node/Ledger.d [hunk 11]
```diff
@@ -247,7 +250,8 @@ public class Ledger
 
     ***************************************************************************/
 
-    public bool acceptBlock (const ref Block block) @safe
+    public bool acceptBlock (const ref Block block)
+        @safe
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Unrelated change.

File: source/agora/node/Ledger.d [hunk 12]
```diff
@@ -535,7 +539,8 @@ public class Ledger
 
     ***************************************************************************/
 
-    public string validateBlock (const ref Block block) nothrow @safe
+    public string validateBlock (const ref Block block)
+        nothrow @safe
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: Unrelated change.

File: source/agora/consensus/state/UTXOSet.d [hunk 13]
```diff
@@ -49,18 +49,20 @@ abstract class UTXOCache
...
+        bool slashed = false) @safe
     {
         import std.algorithm : any;
 
         // defaults to next block
         Height unlock_height = Height(height + 1);
 
         // for payments of frozen transactions, it will melt after 2016 blocks
-        if ((tx.type == TxType.Payment)
+        if ((tx.type == TxType.Payment && !slashed)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/state/UTXOSet.d:68]: You need to document these kinds of changes, it's very important.

For example:

```
If this is an unfreeze and slashing transaction, don't melt but instead set the unlock height to the next block as it will be sent to the commons budget.
```

File: source/agora/node/Ledger.d [hunk 14]
```diff
@@ -395,6 +395,45 @@ public class Ledger
...
+            auto remain_amount = Amount(utxo_value.output.value);
+            remain_amount.sub(this.slash_man.penalty_amount);
+            Transaction slashing_tx =
+            {
+                TxType.Payment,
+                inputs: [Input(utxo)],
+                outputs: [
+                    Output(this.slash_man.penalty_amount,
+                        this.slash_man.penalty_address),
+                    Output(remain_amount, utxo_value.output.address),
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d:500]: This means if the node gets punished then parts of their funds will be sent to the commons budget, but now the leftover part will be immediately unfrozen. I'm not sure this is the best idea. I would have hoped that the leftover part either remains frozen, or becomes a melting transaction.

But I understand it might be complicated to implement.

How about we leave it as is for now and we file a new issue for improving this later?

File: source/agora/node/Ledger.d [hunk 15]
```diff
@@ -427,6 +466,31 @@ public class Ledger
...
+        Update the validators that are to be slashed
+
+        Params:
+            block = the block to update the Validator set with
+
+    ***************************************************************************/
+
+    protected void updateSlashedValidatorSet (const ref Block block) @safe
+    {
+        if(block.header.height == 0)
```
Comment (AndrejMitrovic) [inline source/agora/node/Ledger.d]: `if(` => `if (`

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
A few comments.

File: source/agora/node/Ledger.d (hunk 14 above)
Comment (linked0) [reply source/agora/node/Ledger.d:500]: > How about we leave it as is for now and we file a new issue for improving this later?

Yes, it could be implemented with a more reasonable policy about the frozen UTXO.

File: source/agora/node/Ledger.d (hunk 10 above)
Comment (linked0) [reply source/agora/node/Ledger.d]: > One change can be made here. Take into account the missing_validators in the ConsensusData, and make sure their preimage is not included when calculating the random seed inside of getRandomSeed.

Right, we should avoid the situation where the preimage arriving late could be used to generate the random seed. I'll fix it.

File: source/agora/utils/PrettyPrinter.d (hunk 8 above)
Comment (linked0) [reply source/agora/utils/PrettyPrinter.d]: > In the future I think we should name this just MV that stands for "missing validators".

I agree with it.

File: source/agora/consensus/state/UTXOSet.d (hunk 13 above)
Comment (linked0) [reply source/agora/consensus/state/UTXOSet.d:68]: > You need to document these kinds of changes, it's very important.
> 
> For example:
> 
> ```
> If this is an unfreeze and slashing transaction, don't melt but instead set the unlock height to the next block as it will be sent to the commons budget.
> ```

Ah, I missed it although it's very important.

File: source/agora/node/Ledger.d (hunk 11 above)
Comment (linked0) [reply source/agora/node/Ledger.d]: Updated.

File: source/agora/node/Ledger.d (hunk 12 above)
Comment (linked0) [reply source/agora/node/Ledger.d]: Updated.

File: source/agora/node/Ledger.d (hunk 15 above)
Comment (linked0) [reply source/agora/node/Ledger.d]: Updated.

File: source/agora/consensus/data/Block.d (hunk 9 above)
Comment (linked0) [reply source/agora/consensus/data/Block.d]: Updated.

File: source/agora/utils/PrettyPrinter.d [hunk 16]
```diff
@@ -331,12 +331,15 @@ private struct BlockHeaderFmt
     {
         try
         {
-            formattedWrite(sink, "Height: %d, Prev: %s, Root: %s, Enrollments: [%s]\nSignature: %s,\nValidators: %s",
+            formattedWrite(sink, "Height: %d, Prev: %s, Root: %s, Enrollments: [%s]\nSignature: %s,\nValidators: %s,\nRandomSeed: [%s],\nMPV: [%s]",
```
Comment (hewison-chris) [inline source/agora/utils/PrettyPrinter.d]: Are we going to stick with MPV? I think `missing` or `slashed` might be better.

File: source/agora/test/SlashingMisbehavingValidator.d [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+
+    // block 1
+    txs.each!(tx => nodes[0].putTransaction(tx));
+    network.expectBlock(Height(1));
+
+    // block 2
+    txs = txs.map!(tx => TxBuilder(tx).sign()).array();
+    txs.each!(tx => nodes[0].putTransaction(tx));
+    network.expectBlock(Height(2));
+    auto block2 = nodes[0].getBlocksFrom(2, 2)[0];
```
Comment (hewison-chris) [inline source/agora/test/SlashingMisbehavingValidator.d]: Why up to 2 blocks in getBlockFrom?

**Review by hewison-chris — APPROVED** [review]
Only minor comments.

File: source/agora/test/SlashingMisbehavingValidator.d (hunk 17 above)
Comment (linked0) [reply source/agora/test/SlashingMisbehavingValidator.d]: There's no special meaning in fact. I'll fix it.

File: source/agora/utils/PrettyPrinter.d (hunk 16 above)
Comment (linked0) [reply source/agora/utils/PrettyPrinter.d]: @AndrejMitrovic gave the same opinion about that. I will change it to `Missing Validators` or `MV`.

Comment (linked0) [reply source/agora/utils/PrettyPrinter.d]: Do you think `Missing` is better?

**Comment by linked0** [comment]
All green in the end.
Ready for review.

File: source/agora/utils/PrettyPrinter.d (hunk 16 above)
Comment (hewison-chris) [reply source/agora/utils/PrettyPrinter.d]: In fact `Missing validators` is better.

Comment (AndrejMitrovic) [reply source/agora/utils/PrettyPrinter.d]: Slashed validators is better I think.

"Missing" could sound like they just haven't signed the block yet.

**Comment by AndrejMitrovic** [comment]
@linked0 you also need to add block validation of the random seed in the [validation routine](https://github.com/bpfkorea/agora/blob/b939c3b12798c4b876b621c012aebf9e19a20f02/source/agora/consensus/validation/Block.d#L94).

Do you want to do it in a separate PR?

**Comment by linked0** [comment]
> @linked0 you also need to add block validation of the random seed in the [validation routine](https://github.com/bpfkorea/agora/blob/b939c3b12798c4b876b621c012aebf9e19a20f02/source/agora/consensus/validation/Block.d#L94).
> 
> Do you want to do it in a separate PR?

I'd like to do it in issue #1444 because it needs to modify some routine generating blocks and other related code so it will take some time.

**Review by AndrejMitrovic — APPROVED** [review]
LGTM. I think just the logging could be changed as mentioned.

**Comment by linked0** [comment]
> LGTM. I think just the logging could be changed as mentioned.

I will do it now, thanks.

File: source/agora/utils/PrettyPrinter.d (hunk 16 above)
Comment (linked0) [reply source/agora/utils/PrettyPrinter.d]: Updated to 'Slashed validators`.

File: source/agora/test/SlashingMisbehavingValidator.d (hunk 17 above)
Comment (linked0) [reply source/agora/test/SlashingMisbehavingValidator.d]: Updated.

**Comment by AndrejMitrovic** [comment]
Thank you @linked0!
